RDGSlice: out of core properties

scober opened the pull request:
`RDGSlice` can now load and remove property columns on demand. The interface for this is closer to the `RDGSlice` interface for loading and removing metadata arrays than it is to the `RDG` interface for loading and unloading properties. That is because `RDGSlice` is read-only and provides only a small subset of the interface provided by `RDG`.

Commit:
RDGSlice: Factor out demand loading logic

Commit:
RDGSlice: demand loading of properties

Add demand loading of properties to RDGSlice in a way that mirrors the
demand loading of metadata arrays already present.

Also remove an unnecessary const.

pyrito commented on libtsuba/include/tsuba/RDGSlice.h:
```diff
@@ -74,6 +74,32 @@ class KATANA_EXPORT RDGSlice {
+  // NB: The following interface differs from the one provided by RDG for
+  // property loading/unloading.
+  // For example, this interface uses "remove" instead of "unload" to emphasize
+  // that property modifications will not be persisted (RDGSlice is read-only).
```
This sounds confusing to me. "remove" vs. "unload" don't seem to make the distinction that the modifications will/won't be persisted.

witchel replied:
I am also confused.  Does this delete the property?  It only discards the modifications?

scober replied:
"remove" vs. "unload" is the distinction that `RDG` currently makes. Remove means "remove from memory + assert not modified", unload means "if necessary, write back update column and update in-memory data structures to point to the new file + remove from memory". I can add comments but I would like to maintain the same verbiage as `RDG`.

scober replied:
I added more comments to clarify the intention of the API. But while I was digging around it looked like maybe the `RDG` remove is actually stronger than I thought? It looks like it actually removes the property from the graph, not just from memory. And those are not the semantics of this remove. So maybe we should call these functions unload and the write-only-ness of the interface can just be the lack of an upsert?
UPDATE: I changed the interface to what I described above. Hopefully it is more clear.

pyrito commented on libtsuba/src/RDGSlice.cpp:
```diff
@@ -11,6 +11,142 @@
+
+  if (it == core->part_header().part_prop_info_list().end()) {
+    // caller's function for emptying the array we care about
+    empty();
```
I think this should be documented a bit more to explain why the empty function needs to be passed in. I see that this function is used internally mainly, but still worth a bit more explanation since the burden is placed on the caller.

scober replied:
I added some documentation, is it more clear now?

pyrito replied:
Yes it is much clearer now.

pyrito commented on libtsuba/src/RDGSlice.cpp:
```diff
@@ -11,6 +11,142 @@
+    return tsuba::ErrorCode::PropertyNotFound;
+  }
+
+  std::vector<tsuba::PropStorageInfo*> property{&(*it)};
```
Interesing. So you're initializing a new vector with an iterator. We can do that?

pyrito replied:
Why are you creating a vector here though? Is that the requirement for `AddPropertySlice`?

witchel replied:
It looks to me like he is bracket initializing the vector with a single element that the iterator points to.

pyrito replied:
Yeah I see that now. At a first glance I thought he was initializing a vector based on an iterator that pointed to the beginning of some container and then it just used that iterator to create the rest of the new vector. I see that it's just a single element now.

e-mcginnis replied:
Nit
While I think this is kind of elegant, it requires a bit more thinking than I'd prefer. 
Maybe to add a bit of clarity split this into:
```
tsuba::PropStorageInfo* prop = &(*it);
std::vector<tsuba::PropStorageInfo*> property{prop};
```
The compiler optimizes it all the same according to godbolt.

scober replied:
This has been updated so that there is no longer an iterator at all (searching has been factored out into a `RDGCore` method).

witchel commented on libtsuba/include/tsuba/RDGSlice.h:
```diff
@@ -74,6 +74,32 @@ class KATANA_EXPORT RDGSlice {
+  // Other differences are more subtle - you can't remove a property by its
+  // index in the loaded property table and you can't specify an index for a
+  // loaded property in that table. These differences are because RDGSlice has
+  // many fewer users than RDG and those users require a less rich interface.
```
and because the index interface is bad.

scober replied:
lol. I did wonder why there was an index-based interface.

witchel commented on libtsuba/src/RDGSlice.cpp:
```diff
@@ -11,6 +11,142 @@
 
+namespace {
+katana::Result<void>
+demand_load_metadata_array(
```
What does demand mean here?

scober replied:
"on-demand" - as in "when the user requests it" as opposed to "when the object is constructed".

scober replied:
I changed this name because it seemed like "demand" was causing more confusion than clarity.

witchel commented on libtsuba/src/RDGSlice.cpp:
```diff
@@ -11,6 +11,142 @@
+  auto it = std::find_if(
+      core->part_header().part_prop_info_list().begin(),
+      core->part_header().part_prop_info_list().end(),
+      [&name](tsuba::PropStorageInfo& psi) { return psi.name() == name; });
```
I'm sort of musing here, but I've seen this pattern of iterating through `part_header().part_prop_info_list()` a bunch.  Should there be a `FilterProperties` function in RDGCore that takes in a function that returns a bool and does this traversal and returns matching `PropStorageInfo`s?

scober replied:
I think we could even offer a `SearchForPropInfo(std::string, std::vector<PartPropInfo>)`. I think searching for a single `PartPropInfo` by name is the most common case.

witchel commented on libtsuba/src/RDGSlice.cpp:
```diff
@@ -11,6 +11,142 @@
+                                            ? core->node_properties()
+                                            : core->edge_properties();
+  std::vector<int> indices = table->schema()->GetAllFieldIndices(name);
+  // only one property per name
```
This seems like a weird way to do it.  We guarantee that there are no duplicate property names (within nodes, within edges).  Why not just use `GetFieldByName`?

scober replied:
I could do that. I was looking at the `arrow::Table` api and saw an opportunity to be defensive/assert an invariant. I can drop it.

witchel commented on libtsuba/src/RDGSlice.cpp:
```diff
@@ -289,80 +425,79 @@ tsuba::RDGSlice::local_to_global_id() const {
-      "populating local to global id");
+  KATANA_CHECKED(demand_load_metadata_array(
+      RDGCore::kLocalToUserIDPropName,
+      [&]() {
+        core_->set_local_to_user_id(
+            katana::NullChunkedArray(arrow::uint64(), 0));
+      },
```
Oh, this is the empty array?  Seems like we are playing fast and loose with types, would arrow::null() also work?  If so, do we need to pass in a lambda?

scober replied:
I think it is important that it be an empty array of `arrow::uint64()`.

witchel replied:
Ok, but weird to me to require a lambda that always returns the same thing.  Maybe it is only the same in the code for this PR.

scober replied:
There are other metadata arrays that could make use of this function to load and unload on demand. But the other issue is that these lambdas both use the `core_->set_XXX()` interface. And there is no real connection between the string name of the local to global id array and the function `set_local_to_global_id()`. So we have to pass something to the function letting it know not only what the empty version of the array is, but how to set it.

witchel reviewed (commented):
Looks good, but questions linger.

Commit:
RDGCore: PropStorageInfo finders

A common pattern is to search a RDGCore's PropStorageInfo vectors for a
property with a certain name. As a small quality-of-life improvement,
provide this functionality from RDGCore directly to reduce the number
of times the searching boilerplate has to be written.

e-mcginnis reviewed (approved):
appreciate the clarification on the unload API. 
LGTM

pyrito reviewed (approved):
Merci beaucoup, Simon! Tout semble plus propre!
